data-pipeline/src/pipeline/main.py

```diff
@@ -452,20 +472,129 @@ def run_compute_rag(set_type, year):
     return time_taken
 
 
-def handle_msg(worker_client, msg, worker_queue, complete_queue):
-    msg_payload = json.loads(msg.content)
-    try:
-        msg_payload["pre_process_duration"] = pre_process_data(
-            worker_client, msg_payload["type"], msg_payload["year"]
-        )
+def run_user_defined_rag(
+    year: int,
+    run_id: str,
+    target_urn: str,
+    comparator_set: pd.DataFrame,
+):
+    """
+    Perform user-defined RAG calculations.
+
+    Use the pre-processed "all-schools" data to guarantee coverage of
+    the user-defined comparator-set.
 
-        msg_payload["comparator_set_duration"] = compute_comparator_sets(
-            msg_payload["type"], msg_payload["year"]
+    Note: `SetType` is _always_ "mixed" when persisted, for
+    user-defined comparator-sets; `run_type` is _always_ "default" for
+    the same.
+
+    :param year: financial year in question
+    :param run_id: unique run identifier
+    :param target_urn: URN of the "target" org.
+    :param comparator_set: user-defined comparator-set
+    :return: duration of RAG calculations
+    """
+    set_type = "mixed"
+    run_type = "default"
+
+    all_schools = prepare_data(
+        pd.read_parquet(
+            get_blob("pre-processed", f"{run_type}/{year}/all_schools.parquet")
         )
+    )
 
-        msg_payload["rag_duration"] = run_compute_rag(
-            msg_payload["type"], msg_payload["year"]
+    st = time.time()
+    logger.info(f"Computing user-defined RAG ({run_id}).")
+    df = pd.DataFrame(
+        compute_user_defined_rag(
+            data=all_schools,
+            target_urn=target_urn,
+            set_urns=comparator_set,
         )
+    ).set_index("URN")
+    logger.info(
+        f"Computing user-defined ({run_id}) RAG. Done in {time.time() - st:.2f} seconds."
+    )
+
+    write_blob(
+        "metric-rag",
+        f"{run_type}/{run_id}/user_defined.parquet",
+        df.to_parquet(),
+    )
+
+    insert_metric_rag(
+        run_type=run_type,
+        set_type=set_type,
+        year=run_id,
+        df=df,
+    )
+
+
+def handle_msg(
+    worker_client: Client,
+    msg: QueueMessage,
+    worker_queue: QueueClient,
+    complete_queue: QueueClient,
+):
+    """
+    Process an incoming message.
+
+    TODO: standard message format?
+
+    User-defined comparator set message _content_ will be of the form:
+
+    ```json
+    {
+        "jobId": "24463424-9642-4314-bb55-45424af6e812",
+        "type": "comparator-set",
+        "runId": "c321ef6a-3b1c-4ce2-8e32-0d0167bf2fa7",
+        "year": 2022,
+        "urn": "106057",
```

Comment:
@jrabbott, should this (and the `set` values) be integers when passed here?

Looking at the parquet files in the end-to-end tests, they appear to be parsed as such—not sure if that's intended though or Pandas is being "helpful"…?

Reply:
In the front-end service & APIs URNs are strings rather than ints, so we don't have any issues with leading zeros (same with company number).

Reply:
@jrabbott, any objections if I update my code to assume integers for now? If we've seen the existing behaviour work, I'd prefer to update in line with that (and fix that in isolation if it turns out to be an issue).

Reply:
No objections from me.